### PR TITLE
Fix and test the full config with gcc and clang

### LINKED ISSUE #2851: Build warning in ssl_server2 in the full config with gcc -Os
### Description
- Type: Bug
- Priority: Major

---------------------------------------------------------------
## Bug

* Environment: Ubuntu 16.04, gcc 5.4.0
* Library: commit c835672c51652586e815c8723335f17a2641eb9e (`development` as of right now, which is [`mbedtls-2.19.1`](https://github.com/ARMmbed/mbedtls/releases/tag/mbedtls-2.19.1))

Bug: failed build (warning) with some build options  
Workaround: ignore the warning or use different build options  

```
./scripts/config.pl full
make CFLAGS='-Wall -Wextra -Werror -Os' lib programs
…
ssl/ssl_server2.c: In function ‘main’:
ssl/ssl_server2.c:4274:9: error: ‘context_buf_len’ may be used uninitialized in this function [-Werror=maybe-uninitialized]
         mbedtls_platform_zeroize( context_buf, context_buf_len );
         ^
cc1: all warnings being treated as errors
Makefile:243: recipe for target 'ssl/ssl_server2' failed
```

The build passes in the default config, or with `gcc -O2`, or with Clang 3.8.0 or 6.0.0. The build fails with the same error with GCC 7.4.0.

### PR DESCRIPTION
**Pull request by gilles-peskine-arm**
Fix #2851. The warning pointed to a genuine bug, but only in a corner case in a test program with a non-default build option.

Expand `all.sh` to ensure that similar bugs would be caught.

Backports: the buggy code didn't exist in the LTS branches, but I think the `all.sh` changes should be backported. [edited by mpg: #3029 #3030]

### CODE REVIEW

File: tests/scripts/all.sh [hunk 1]
```diff
@@ -954,8 +954,8 @@ component_test_no_platform () {
     scripts/config.pl unset MBEDTLS_PSA_ITS_FILE_C
     # Note, _DEFAULT_SOURCE needs to be defined for platforms using glibc version >2.19,
     # to re-enable platform integration features otherwise disabled in C99 builds
-    make CC=gcc CFLAGS='-Werror -Wall -Wextra -std=c99 -pedantic -O0 -D_DEFAULT_SOURCE' lib programs
-    make CC=gcc CFLAGS='-Werror -Wall -Wextra -O0' test
+    make CC=gcc CFLAGS='-Werror -Wall -Wextra -std=c99 -pedantic -Os -D_DEFAULT_SOURCE' lib programs
```
[the remark this answers is not in the document]
Comment (gilles-peskine-arm) [reply tests/scripts/all.sh]: I explained why I wanted to avoid `-O0` builds in the commit message. I didn't have any particular reason to pick `-Os` vs `-O1`.

A reason not to pick `-Olarger` is that the compilation is slightly slower, OTOH a reason to pick `-Olarger` is that the tests run slightly faster. But the difference is tiny.

[the remark this answers is not in the document]
Comment (AndrzejKurek) [reply tests/scripts/all.sh]: Which analyses or measurements exactly are omitted when compiling with `-O0`?

[the remark this answers is not in the document]
Comment (gilles-peskine-arm) [reply tests/scripts/all.sh]: I don't know the exact list. I suspect you'd need to look at the source of individual compilers. There's some information in the [documentation of GCC warnings](https://gcc.gnu.org/onlinedocs/gcc-8.3.0/gcc/Warning-Options.html): search for `optim` to see that some warnings are less effective or disabled at `-O0`. `-Wmaybe-uninitialized` is a pretty important one that's disabled with `gcc -O0`.

**Comment by AndrzejKurek** [comment]
I'm fine with the changes, but there's a conflict in all.sh that needs to be resolved first.

File: tests/scripts/all.sh [hunk 2]
```diff
@@ -964,21 +964,21 @@ component_build_no_std_function () {
...
-    make CC=gcc CFLAGS='-Werror -Wall -Wextra -O0'
+    make CC=gcc CFLAGS='-Werror -Wall -Wextra -Os'
 }
 
 component_build_no_ssl_srv () {
     msg "build: full config except ssl_srv.c, make, gcc" # ~ 30s
     scripts/config.pl full
     scripts/config.pl unset MBEDTLS_SSL_SRV_C
-    make CC=gcc CFLAGS='-Werror -Wall -Wextra -O0'
+    make CC=gcc CFLAGS='-Werror -Wall -Wextra -O1'
```
Comment (piotr-now) [inline tests/scripts/all.sh]: I observe the same warning as in mentioned #2851 only in case of `-Os` and `-Og`. Why did you chose `-O1` in some of components? From GCC documentation I would assumed that the more optimization, the better chance of warnings detection.

File: tests/scripts/all.sh [hunk 3]
```diff
@@ -1113,6 +1113,30 @@ component_test_cmake_shared () {
...
+
+component_test_clang_opt () {
+    scripts/config.pl full
+    test_build_opt 'full config' clang -O0 -Os -O2
+}
+
+component_test_gcc_opt () {
+    scripts/config.pl full
+    test_build_opt 'full config' gcc -O0 -Os -O2
+}
```
Comment (piotr-now) [inline tests/scripts/all.sh]: What do you think about change `-O2` in to `-O3`? I think `-O3` is more common option

File: tests/scripts/all.sh (hunk 2 above)
Comment (gilles-peskine-arm) [reply tests/scripts/all.sh]: Most warnings that depend on optimization are enabled if any optimization is enabled. A few cases are only triggered at large optimization, but that's rare in my limited experience. I generally chose `-O1` rather than `-O2` or more when building something that won't be executed because `-O1` is faster and `-O2` or more gives a very limited return on investment.

File: tests/scripts/all.sh (hunk 3 above)
Comment (gilles-peskine-arm) [reply tests/scripts/all.sh]: In my experience `-O2` is more common and larger levels tend to invite compiler bugs, but this perception may be wrong or outdated.

**Comment by gilles-peskine-arm** [comment]
Rebased :frowning_face: and backported.
